src/strtoul.rs

```diff
@@ -1,67 +1,188 @@
+//! OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+//! HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+//! LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+//! OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+//! SUCH DAMAGE.
+//!
+//! Translated from https://github.com/gcc-mirror/gcc/blob/97d1ed67fc6a5773c8c00875bfa3616a457cf5f9/libiberty/strtoul.c
 
-use crate::{CChar, CStringIter, CULong};
+use crate::{CChar, CInt, CLong, CULong};
 
-/// Rust implementation of C library function `strtoul`.
+/// Rust implementation of C library function [`strtoul`](https://cplusplus.com/reference/cstdlib/strtoul/).
+///
+/// Passing NULL (core::ptr::null()) gives undefined behaviour.
 ///
-/// Takes a null-terminated string and interprets it as a decimal integer.
-/// This integer is returned as a `CULong`. Parsing stops when the first
-/// non-digit ASCII byte is seen. If no valid ASCII digit bytes are seen, this
-/// function returns zero.
+/// Convert a string to an unsigned long integer.
+///
+/// Ignores `locale' stuff.  Assumes that the upper and lower case
+/// alphabets and digits are each contiguous.
 #[no_mangle]
-pub unsafe extern "C" fn strtoul(s: *const CChar) -> CULong {
-	let mut result: CULong = 0;
-	for c in CStringIter::new(s) {
-		if c >= b'0' && c <= b'9' {
-			result *= 10;
-			result += (c - b'0') as CULong;
+pub unsafe extern "C" fn strtoul(
+	nptr: *const CChar,
+	endptr: *mut *const CChar,
+	mut base: CInt,
+) -> CULong {
+	let mut s = nptr;
+
+	let mut c = *s;
+	s = s.offset(1);
+	while isspace(c) {
+		c = *s;
+		s = s.offset(1);
+	}
+
+	let neg = if c == b'-' {
+		c = *s;
+		s = s.offset(1);
+		true
+	} else {
+		if c == b'+' {
+			c = *s;
+			s = s.offset(1);
+		}
+		false
+	};
+
+	if (base == 0 || base == 16) && c == b'0' && (*s == b'x' || *s == b'X') {
+		c = *s.offset(1);
+		s = s.offset(2);
+		base = 16;
+	}
+
+	if base == 0 {
+		base = if c == b'0' { 8 } else { 10 };
+	}
+
+	let cutoff = CULong::MAX / base as CULong;
+	let cutlim = CULong::MAX % base as CULong;
+
+	let mut acc = 0;
+	let mut any = 0;
+
+	loop {
+		if isdigit(c) {
+			c -= b'0';
+		} else if isalpha(c) {
+			c -= if isupper(c) { b'A' - 10 } else { b'a' - 10 };
 		} else {
 			break;
 		}
+
+		if c as CInt >= base {
+			break;
+		}
+
+		if any < 0 {
+			c = *s;
+			s = s.offset(1);
+			continue;
+		}
+
+		if acc > cutoff || (acc == cutoff && c as CULong > cutlim) {
+			any = -1;
+			acc = CULong::MAX;
+		} else {
+			any = 1;
+			acc *= base as CULong;
+			acc += c as CULong;
+		}
+
+		c = *s;
+		s = s.offset(1);
 	}
-	result
+	if neg && any > 0 {
+		acc = -(acc as CLong) as _;
+	}
+
+	if !endptr.is_null() {
+		(*endptr) = if any != 0 {
+			s.offset(-1)
+		} else {
+			core::ptr::null()
+		};
+	}
+
+	acc
+}
+
+fn isspace(argument: CChar) -> bool {
+	const SPACE_CHARACTERS: [u8; 6] = [b' ', b'\n', b'\t', b'v', b'f', b'r'];
```

Comment:
Are v, f and r space characters?

Reply:
According to C they are: https://cplusplus.com/reference/cctype/isspace/

Reply:
Not without the backslashes :)

Reply:
Oh! Totally missed that! Good catch

Reply:
Ok, fixed it. Apparently Rust doesn't support all C escapes